build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1288,6 +1670,228 @@ message WaitExecutionRequest {
+// two-phase cache lookups.
+message GetPotentialActionResultsRequest {
+  // The instance of the execution system to operate against. A server may
+  // support multiple instances of the execution system (with their own workers,
+  // storage, caches, etc.). The server MAY require use of this field to select
+  // between them in an implementation-defined fashion, otherwise it can be
+  // omitted.
+  string instance_name = 1;
+
+  // The hash of an [Action][build.bazel.remote.execution.v2.Action] that may
+  // contain incomplete file input predictions. This possibly-incomplete `Action`
+  // acts as a key to look up potential action results that must be compared
+  // to the client's current filesystem contents to get a match. On a match,
+  // the corresponding build outputs in the form of an
+  // [ActionResult][build.bazel.remote.execution.v2.ActionResult] can be
+  // retrieved from
+  // [ActionCache.GetActionResult][build.bazel.remote.execution.v2.ActionCache.GetActionResult].
+  //
+  // In two-phase caching an `Action` contains a reference to a
+  // [Command][build.bazel.remote.execution.v2.Command] but the
+  // `input_root_digest` can refer to a subset of the files that are predicted
+  // to be needed to run the `Action`. The input root can contain all files
+  // needed for the command, in which case the action fingerprint should lead
+  // to just one output (or none if there are no cache hits). When the input
+  // root contains fewer files than those actually needed by the action, the
+  // action fingerprint gets less specific, and the number of possible matches
+  // grows. It is possible for the input root to contain zero files, which is
+  // a very non-specific fingerprint as it can match any action results that
+  // match the `Command`. Less-specific action fingerprints lead to more I/O
+  // and time spent matching `ObservedPathSet`s against the client filesystem;
+  // adding in any predictions about the filesystem "focuses" the lookup to
+  // more-specific results.
+  Digest action_fingerprint_digest = 2;
+
+  // A hint to the server to request inlining `ObservedPathSet`s in the
+  // [GetPotentialActionResultsResponse][build.bazel.remote.execution.v2.GetPotentialActionResultsResponse]
+  // message.
+  bool inline_pathsets = 3;
+
+  // A maximum page size to request. If present, the server will respond with no more
+  // than this many items. Regardless of whether a page size is specified, the
+  // server may place its own limit on the number of items to be returned and
+  // require the client to retrieve more items using a subsequent request.
+  int32 page_size = 4;
+
+  // A page token, which must be a value received in a previous
+  // [GetPotentialActionResultsResponse][build.bazel.remote.execution.v2.GetPotentialActionResultsResponse].
+  // If present, the server will use that token as an offset, returning only
+  // that page and the ones that succeed it.
+  string page_token = 5;
```

Comment:
What if an action has, say, thousands of potential action results? Is it still worth doing action cache lookups at that point?

More general question: is it still worth in practice to even cache actions that are known not to be hermetic? Doesn't the bandwidth/computation requirements at some point exceed the amount of work that's done by the action?

Reply:
Speaking from experience direct and indirect - when you end up with thousands someone is doing something wrong or interesting. We have had such live-site issues and the fixes were varied, either by focusing the action_fingerprint_digest key by adding more specific input hashes to get fewer results, or letting the client decide when it is done reviewing results e.g. as a time-boxed effort. I was not aiming with the doc comments here to force any specific implementation since the user intent can vary (maybe like one case they actually wanted to spend the time on the hundreds of potential results).